File: org.opencompare/io-best-buy/src/main/java/impl/BestBuyMiner.java
package impl;

import domain.BestBuyProduct;
import java.util.List;
import org.opencompare.api.java.Cell;
import org.opencompare.api.java.Feature;
import org.opencompare.api.java.PCM;
import org.opencompare.api.java.PCMFactory;
import org.opencompare.api.java.Product;
import org.opencompare.api.java.impl.PCMFactoryImpl;

/**
 * BestBuyMiner provides methods to manipulate and analyze bestbuy products.
 * 
 * @author jmdavril
 */
public class BestBuyMiner {

    private final PCMFactory pcmFactory;

    public BestBuyMiner() {
        this.pcmFactory = new PCMFactoryImpl();
    }

    /**
     * @requires products != null
     * @return the PCM corresponding to the given list of bestbuy products
     */
    public PCM createPCM(List<BestBuyProduct> products) {
        assert (products != null);
        for (BestBuyProduct p : products) {
            assert (p != null);
        }

        PCM pcm = this.pcmFactory.createPCM();

        Feature productsKey = pcmFactory.createFeature();
        pcm.addFeature(productsKey);
        pcm.setProductsKey(productsKey);

        for (BestBuyProduct p : products) {
            Product product = this.pcmFactory.createProduct();

            Cell productKey = pcmFactory.createCell();
            productKey.setFeature(productsKey);
            productKey.setContent(p.getName());

            pcm.addProduct(product);
            if (p.getDetails() != null) {
                for (BestBuyProduct.Detail d : p.getDetails()) {
                    Feature feature = pcm.getOrCreateFeature(d.getName(), this.pcmFactory);
                    Cell cell = this.pcmFactory.createCell();
                    cell.setContent(d.getValue());
                    cell.setFeature(feature);
                    product.addCell(cell);
                }
            }
        }
        
        pcm.normalize(this.pcmFactory);

        return pcm;
    }
}

File: org.opencompare/io-best-buy/src/main/scala/org/opencompare/io/bestbuy/BestBuyMiner.scala
package org.opencompare.io.bestbuy

import org.opencompare.api.java.{PCM, PCMFactory}

/**
 * Created by gbecan on 09/02/15.
 */
class BestBuyMiner(factory : PCMFactory) {

  private val bestBuyAPI = new BestBuyAPI

  def minePCM(skus : List[String]) : PCM = {
    // Create product, features and cells
    val productInfos = for (sku <- skus) yield {
      bestBuyAPI.getProductInfo(sku)
    }

    mergeSpecifications(productInfos)
  }

  def mergeSpecifications(productInfos : List[ProductInfo]): PCM = {
    val pcm = factory.createPCM()

    // Create products key feature
    val skuFeature = factory.createFeature()
    pcm.addFeature(skuFeature)
    pcm.setProductsKey(skuFeature)

    for (productInfo <- productInfos) {
      val product = factory.createProduct()
      pcm.addProduct(product)

      val productName = productInfo.sku
      val skuCell = factory.createCell()
      skuCell.setFeature(skuFeature)
      skuCell.setContent(productName)

      for ((featureName, value) <- productInfo.details) {
        val feature = pcm.getOrCreateFeature(featureName, factory)
        val cell = factory.createCell()
        cell.setContent(value)
        cell.setFeature(feature)
        product.addCell(cell)
      }
    }

    pcm.normalize(factory)

    pcm
  }

}
